components/support/sql/src/open_database.rs

```diff
@@ -35,89 +42,106 @@ pub enum Error {
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-pub trait MigrationLogic {
+pub trait ConnectionInitializer {
     // Name to display in the logs
     const NAME: &'static str;
 
     // The version that the last upgrade function upgrades to.
     const END_VERSION: u32;
 
-    // Runs first, outside of a transaction
-    fn setup_pragmas(&self, _conn: &Connection) -> Result<()> {
-        Ok(())
-    }
+    // Functions called only for writable connections all take a Transaction
+    // Initialize a newly created database to END_VERSION
+    fn init(&self, tx: &Transaction<'_>) -> Result<()>;
 
-    // Runs before the init/upgrade functions
+    // Upgrade schema from version -> version + 1
+    fn upgrade_from(&self, conn: &Transaction<'_>, version: u32) -> Result<()>;
+
+    // Runs immediately after creation for all types of connections. If writable,
+    // will *not* be in the transaction created for the "only writable" functions above.
     fn prepare(&self, _conn: &Connection) -> Result<()> {
         Ok(())
     }
 
-    // Initialize a newly created database to END_VERSION
-    fn init(&self, conn: &Connection) -> Result<()>;
-
-    // Upgrade schema from version -> version + 1
-    fn upgrade_from(&self, conn: &Connection, version: u32) -> Result<()>;
-
-    // Runs after the init/upgrade functions
+    // Runs for all types of connections. If a writable connection is being
+    // initialized, this will be called after all initialization functions,
+    // but inside their transaction.
```

Comment:
Combining `prepare` and `setup_pragmas` makes a lot of sense.  `setup_pragmas` always felt overly specific me.  Also +1 to explicitly stating which methods are inside a transaction vs a connection.